[Property Wrappers] Fix property wrapper initialization type checking when `wrappedValue` is an optional of a generic parameter.

hborla opened the pull request:
* Make `repairViaOptionalUnwrap` less eager
* Don't anchor the equality constraint between the property type and `wrappedValue` type at the initialization expression. Failure to solve this constraint should be diagnosed at the wrapped property itself 

Resolves: rdar://problem/65802331

hborla commented:
@swift-ci please smoke test

hborla commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2866,7 +2866,7 @@ repairViaOptionalUnwrap(ConstraintSystem &cs, Type fromType, Type toType,
   fromType = fromType->getWithoutSpecifierType();
 
-  if (!fromType->getOptionalObjectType() || toType->is<TypeVariableType>())
+  if (!fromType->getOptionalObjectType() || toType->hasTypeVariable())
```
The case that I'm trying to fix is exactly the case you describe. `matchTypes` is matching `Optional<Optional<Int>>` and `Optional<$T0>` and in here it's binding `$T0` to `Int` directly. I don't think this makes sense to do from `repairFailures` since this is essentially just collecting potential fixes to try which can become terms in a disjunction (I know that it's usually only one fix, but the fix can be in a disjunction with other conversions) - I think it's really unexpected that `repairViaOptionalUnwrap` would bind a type variable before attempting the fix. If we do want to look through optional types to match the context with the unwrapped type, why wouldn't we want to do that via potential binding like we do for other fixes that bind type variables?

xedin replied:
> If we do want to look through optional types to match the context with the unwrapped type, why wouldn't we want to do that via potential binding like we do for other fixes that bind type variables? Why wouldn't we want to do that for plain type variables too?

I think the challenge would be exactly the same, we'd need to know when it's acceptable to infer something with N levels of optional stripped off and it would mean that "fix" logic would have to be introduced into binding inference as well..

xedin committed:
Make `repairViaOptionalUnwrap` less eager

If the source of the "mismatch" is `Equal` constraint we can't
record a missing unwrap fix if either size is a type variable
after all of the optionals have been stripped away because
there is not enough information about what is expected behavior
until type variable gets bound.

hborla committed:
[Property Wrappers] Always anchor the equality constraint between
the property type and wrapped value type at the wrapped VarDecl.

hborla commented:
@swift-ci please smoke test

xedin reviewed (approved):
LGTM!

AnthonyLatsis commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2958,6 +2958,19 @@ repairViaOptionalUnwrap(ConstraintSystem &cs, Type fromType, Type toType,
   std::tie(fromObjectType, fromUnwraps) = getObjectTypeAndUnwraps(fromType);
   std::tie(toObjectType, toUnwraps) = getObjectTypeAndUnwraps(toType);
 
+  // Since equality is symmetric and it decays into a `Bind` eagerly
```
I suppose there's a missing comma here before `eagerly`.

hborla replied:
Fixed, thank you!

hborla committed:
[NFC] Fix grammar in `repairViaOptionalUnwrap` comment

hborla commented:
@swift-ci please smoke test
